### ISSUE TITLE
Group refactor

### ISSUE THREAD

**Issue by Psirus** [nutofem/nuto#63]
The groups are another one of these areas that could simply be nicer.

- they don't need to be stored at the structure
- they don't need to have member IDs
- there is no need for a base class
- the operations `Difference`, `Intersection` and so on should be free functions

Oh, and things like `GetType` or `GetTypeString` should be a punishable offense anyway.

In addition, all the group related `Structure` methods should either go away to where they came from, or be converted to free functions too.

What has been stopping me (apart from lack of time) is that I would like to iterate over the group, i.e.
~~~cpp
for(auto& member : group)
    member.DoMemberishThings();
~~~

What I've been able to come up with so far can be seen in commit f433b77. There, the group saves raw pointers to its constituents and the `begin()` and `end()` iterators have been implemented as `boost::indirect_iterator`s to get the value semantics.
Pro:
- `auto`
- value semantics

Con:
- raw ptr
- `begin` and `end` behave differently from normal

Another option would be to use `std::reference_wrapper` instead of raw pointers. This would get you the value semantics and normally behaving `begin` and `end`, but no `auto`.

Maybe you have an idea that circumvents all the cons? Shark below :shark:

**Comment by TTitscher** [comment]
IMO a raw ptr is nothing bad. In contrast to a raw _owing_ ptr that requires `new` and `delete`. So if the ptr is hidden as an implementation detail and accessed via reference, I see no con there.

**Comment by Psirus** [comment]
Well, sort of. What happens if the object pointed to goes out of scope earlier than the group? Unlikely in our use cases, but still.

**Comment by joergfunger** [comment]
Is it that unlikely that the object goes out of scope? IMO, we do sometimes a reallocation of certain components (i.e. currently nodes when the dofs are changing, static data when the integration type (and thus the number of static data changes), .... If the group is just stored outside (of the structure) either with a raw ptr or a shared ptr, this might cause issues. (either the object is out of scope and the ptr is not valid, or the object is not deleted but now the group object and the structure object are different). This would require some exchange_ptr methods.
Similar, the restore procedure for the structure (before starting e.g. the visualization/restart), this changes the pointers as well, so all groups would have to be recreated.

**Comment by TTitscher** [comment]
The question, if we use a ptr/`std::reference_wrapper`/whatever does not relate to the _scope_. And IMO we should not overthink this scope part. We only use `Group<Node>` and `Group<Element>`.  If our elements and nodes somehow go out of scope, we cannot do FEM anymore. So why should they? And nodes only change internally in `ConvertToInterpolationType`. This will be gone with separated dofs. 

So some other questions:
- On a scale from 0 to 9, how important are value semantics?
- Should the group work with algorithms like `std::sort`?
- Duplicates, are they allowed/intended? If not, do we check that or mark it as _undefined behavior_?
- How important is performance?

**Comment by Psirus** [comment]
I agree on not overthinking the scope thing, but concealing it seemed dishonest.

Value semantics are quite important to me, unless we want to name them `PointerGroup`. Good interfaces makes stuff so much easier and enjoyable to use. So on a scale from 0 to 7, I'd say 5.

Don't know on the duplicates. Accidentally adding a boundary condition for a node twice would be bad, but is that the fault of the group for not telling you, or of the user for putting in a node twice?

Performance is not really crucial here. Most of the stuff with groups you do once, like setting boundary conditions or defining visualization output. I'd rather have a nice implementation sooner, and if the performance ever becomes a problem, worry about it then.

**Comment by TTitscher** [comment]
On the duplicates. We could allow arbitrary adding and provide methods to deal with duplicates manually.
~~~
bool ContainsDuplicates()
void RemoveDuplicates()
~~~

**Comment by Psirus** [comment]
The current idea can now be seen in [this Gist](https://gist.github.com/Psirus/b85a51a50d47ae32f24710e2abf5969e), rather than keeping an outdated branch around. We can pick this up later without a huge merge.

**Comment by TTitscher** [comment]
__TL/DR:__ Decoupling of `Group` and `UniqueId` (IMO easily) possible and causes a lot less headaches. 

__Sorry to warm this up again. :ghost:__ 

Currently we provide a deterministic ordering of group members via their Id provided by `NuTo::UniqueId`. It turns out that this causes some problems. They might be solved somehow, but constantly thinking about those issues decelerates the whole development.
- Mocking. This might not be obvious (because lots of nuto developers do not use mocks (yet!)), but is really annoying. If I create two mock objects of type `fakeit::Mock<InterfaceWithUniqueId>`, they both have the same Id (0). So I cannot mock a `NuTo::Group<InterfaceWithUniqueId>` since all have Id==0.
- Copy/Clone. Does the clone of `InterfaceWithUniqueId` have a new Id? What about a copy of an object derived from that?

So I strongly vote for removing the UniqueId requirement in the Groups. UniqueId itself is very useful (mainly debugging) and we can define an arbitrary behavior on copy/clone. But no other features of NuTo should rely on that.

The requirement of the group is to have some kind of deterministic ordering. I propose a _chronological_ ordering. The first entity is and stays at the front of the group. (e.g. `vector::push_back`). This is very close to a proposal of @vhirtham. We did not thought further about his approach because the idea to avoid duplicates was to use the inefficent `std::find` on each insertion. 

How to we efficiently avoid duplicates? 
- Provide a `::RemoveDuplicates()` that is called from time to time
    - manually, if you really care about duplicates _and/or_
    - automatically in algorithms, e.g. before a group is returned _and/or_
    - after algorithms like ::Intersection/::Difference/etc.
- Check for duplicates every time
    - provide an equivalent std::set (sorted by ptr, but we don't care. The std::vector provides our ordering). If set insertion fails (==duplicate) avoid adding to the group. Small memory overhead, my favorite nonetheless.

So again. I'm not saying, we cannot solve the problems above. And actually, I do not want to discuss solutions to the problems above. We can (IMO easily) avoid them totally and not think about them again. __Any objections?__ :smile_cat: 

__Edit:__ _small_ memory overhead actually means double the memory. Just references. Still no big deal IMO.

**Comment by pmueller2** [comment]
Ok, I think: duplicates yes or no. No intermediate Remove from time to time etc. If the memory overhead is so small that nobody cares than go for the combined set and vector version (.. it still feels somehow ... strange).
